plugins/StereoControl/StereoControl.cpp

```diff
@@ -0,0 +1,444 @@
+				multimodeFilter(s[i], multimodeCoeff(monoBassFreq), lp[i], hp[i], m_monoBassFilter[i][0]);
+				multimodeFilter(s[i], multimodeCoeff(monoBassFreq), lp[i], hp[i], m_monoBassFilter[i][1]);
+			}
+
+			float lowmono = (lp[0] + lp[1]) * 0.5f;
+
+			if (audition)
+			{
+				s[0] = lowmono;
+				s[1] = lowmono;
+			}
+			else
+			{
+				s[0] = lowmono - hp[0];// Subtraction is needed due to filter phase shift
+				s[1] = lowmono - hp[1];
+			}
+		}
+
+		switch (panMode)
+		{
+			case 0:// Gain
+			{
+				const float lGain = pan > 0 ? 1.f - pan : 1.f;
+				const float rGain = pan < 0 ? 1.f + pan : 1.f;
+				s[0] *= lGain;
+				s[1] *= rGain;
+				break;
+			}
+			case 1:// Stereo
+			{
+				const float lGainL = panDualL > 0 ? 1.f - panDualL : 1.f;
+				const float rGainL = panDualL < 0 ? 1.f + panDualL : 1.f;
+				const float lGainR = panDualR > 0 ? 1.f - panDualR : 1.f;
+				const float rGainR = panDualR < 0 ? 1.f + panDualR : 1.f;
+				const float temp = s[0];
+				s[0] = s[0] * lGainL + s[1] * lGainR;
+				s[1] = temp * rGainL + s[1] * rGainR;
+				break;
+			}
+			case 2:// Haas
+			{
+				float haasResult = 0;
+				float haasDelayVal[2] = {0};
+
+				if (pan >= 0)
+				{
+					haasDelayVal[0] = linearInterpolate(0, 0.0008f * m_sampleRate * panDelay, abs(pan));
+				}
+				else
+				{
+					haasDelayVal[1] = linearInterpolate(0, 0.0008f * m_sampleRate * panDelay, abs(pan));
+				}
+
+				for (int i = 0; i < 2; ++i)
+				{
+					if ((i == 0 && pan > 0) || (i == 1 && pan < 0))
+					{
+						// Read next value from delay buffer
+						float readLoc = m_haasIndex[i] - haasDelayVal[i];
+						if (readLoc < 0) {readLoc += m_haasBufSize;}
+						float readLocFrac = fraction(readLoc);
+						if (readLoc < m_haasBufSize - 1)
+						{
+							haasResult = linearInterpolate(m_haasBuf[i][floor(readLoc)], m_haasBuf[i][ceil(readLoc)], readLocFrac);
+						}
+						else// For when the interpolation wraps around to the beginning of the buffer
+						{
+							haasResult = linearInterpolate(m_haasBuf[i][m_haasBufSize - 1], m_haasBuf[i][0], readLocFrac);
+						}
+					}
+
+					// Increment delay ring buffer index
+					++m_haasIndex[i];
+					if (m_haasIndex[i] >= m_haasBufSize)
+					{
+						m_haasIndex[i] -= m_haasBufSize;
+					}
+					// Send new value to delay line
+					m_haasBuf[i][m_haasIndex[i]] = s[i];
+				}
+
+				if (pan != 0)
+				{
+					const float minGain = linearInterpolate(1.f, 0.15f, abs(pan));
+					if (pan >= 0)
+					{
+						const float lGain = ((1.f - pan) + minGain) / (1.f + minGain);
+						s[0] = haasResult * lGain;
+					}
+					else
+					{
+						const float rGain = ((1.f + pan) + minGain) / (1.f + minGain);
+						s[1] = haasResult * rGain;
+					}
+				}
+
+				const int lowChnl = (pan <= 0);
+				const int highChnl = !lowChnl;
+
+				float lp;
+				float hp;
+
+				// Filter parameters are semi-arbitrary, but are vaguely based on some random graphs I saw during research
```

Comment:
If you remember the sources, please mention them. Also, how did you get the gain formula?

Reply:
Took me a while to scrape up the sites I read through while making this, but I think this part was mostly inspired by the graphs on this page: https://goodhertz.co/panpot/